### PR TITLE
Metadata updates

### PR DESCRIPTION
**Pull request by max-zilla**
Adds additional metadata features including deleting, PUT and PATCH, more validation on metadata definitions.

### CODE REVIEW

File: backend/app/models/metadata.py [hunk 1]
```diff
@@ -1,67 +1,136 @@
...
+
+def validate_definition(contents: dict, metadata_def: MetadataDefinitionOut):
+    """Convenience function for checking if a value matches MetadataDefinition criteria"""
+    for field in metadata_def.fields:
+        if field.name in contents:
+            t = FIELD_TYPES[field.type]
+            try:
+                contents[field.name] = t(contents[field.name])
+            except ValueError:
+                if field.list and type(value) is list:
```
Comment (longshuicy) [inline backend/app/models/metadata.py:106]: here the `value` is not defined

File: backend/app/routers/metadata_files.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    db: MongoClient = Depends(dependencies.get_db),
+):
+    """Attach new metadata to a file. The body must include a contents field with the JSON metadata, and either a
+    context JSON-LD object, context_url, or definition (name of a metadata definition) to be valid.
+
+    Returns:
+        Metadata document that was added to database
+    """
+    if (file := await db["files"].find_one({"_id": ObjectId(file_id)})) is not None:
+        md = _build_metadata_db_obj(db, metadata_in, FileOut(**file))
```
Comment (longshuicy) [inline backend/app/routers/metadata_files.py:98]: I got an error here 
```
 File "/Users/cwang138/Documents/Clowder/clowder2/backend/app/routers/metadata_files.py", line 98, in add_file_metadata
    md = _build_metadata_db_obj(db, metadata_in, FileOut(**file), user)
TypeError: _build_metadata_db_obj() missing 1 required positional argument: 'user'
```

when trying to attach metadata to a file:
```
{
  "definition": "LatLon",
  "contents": {
  	"longitude":40.123,
  	"latitude":-88.777
  }
}
```

File: backend/app/routers/metadata_datasets.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                md_obj.to_mongo())
+            found = await db["metadata"].find_one({"_id": md["_id"]})
+            metadata_out = MetadataOut.from_mongo(found)
+            return metadata_out
+    else:
+        raise HTTPException(status_code=404, detail=f"Dataset {dataset_id} not found")
+
+
+@router.patch("/{dataset_id}/metadata", response_model=MetadataOut)
+async def update_dataset_metadata(
```
Comment (longshuicy) [inline backend/app/routers/metadata_datasets.py:136]: I'm trying to patch an exist metadata entry by giving the "definition", "metadata id" as well as the "contents". But somehow it's not patching the correct metadata?

My request is
```
{"id":"628664405685a1694705c863","definition":"LatLon","contents":{"latitude":"111","longitude":"222"}}
```

and my response is on another metadata definition:
```
{"id":"6286642c5685a1694705c85f","context":null,"context_url":"clowder.org","definition":null,"contents":{"color":"blue","latitude":"111","longitude":"222"},"resource":{"collection":"datasets","resource_id":"627a8d09ca3d2920a17f602a","version":null},"agent":{"id":"6286642c5685a1694705c860","creator":{"id":"627a8d01ca3d2920a17f6025","email":"cwang138@illinois.edu","first_name":"Chen","last_name":"Wang"},"extractor":null},"created":"2022-05-19T15:37:16.065000"}
```

File: backend/app/models/metadata.py (hunk 1 above)
Comment (max-zilla) [reply backend/app/models/metadata.py:115]: fixed

File: backend/app/routers/metadata_datasets.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        Metadata document that was added to database
+    """
+    if (
+        dataset := await db["datasets"].find_one({"_id": ObjectId(dataset_id)})
+    ) is not None:
+        # If dataset already has metadata using this definition, don't allow duplication
+        definition = metadata_in.definition
+        if definition is not None:
+            existing_q = {
+                "resource.resource_id": dataset.id,
```
Comment (longshuicy) [inline backend/app/routers/metadata_datasets.py:87]: I got an error complaing that dict object doesn't have id field and I looked at the object it has "_id" instead: 
image

I could fix it on my machine but then i looked at similar endpoints on `metadata_file.py` and saw you are using file.id there and it worked... No sure what's going on here? @max-zilla

Comment (longshuicy) [reply backend/app/routers/metadata_datasets.py:87]: Oh actually i know... it's missing `dataset = DatasetOut(**dataset)`.

**Review by longshuicy — APPROVED** [review]

**Pull request merged**